torch/utils/checkpoint.py

```diff
@@ -0,0 +1,118 @@
+
+
+def unpack_variables(inputs):
+    if type(inputs) == Variable:
+        return inputs.data
+    elif torch.is_tensor(inputs):
+        return inputs
+    elif isinstance(inputs, tuple):
+        return tuple(unpack_variables(v) for v in inputs)
+    else:
+        raise RuntimeError("Unknown input type")
+
+
+class CheckpointFunction(Function):
+
+    # NOTE: *args is the flat inputs list, args is the tuple containing inputs
+    @staticmethod
+    def forward(ctx, run_function, *args):
+        ctx.run_function = run_function
+        ctx.save_for_backward(*args)
+        inputs = repackage_inputs(args)
+        with torch.no_grad():
+            outputs = run_function(*inputs)     # the *inputs* is always a tuple
+        return unpack_variables(outputs)
+
+    @staticmethod
+    def backward(ctx, *args):
+        saved_inputs = ctx.saved_tensors
+        inputs = repackage_inputs(saved_inputs)
+        with torch.enable_grad():
+            outputs = ctx.run_function(*inputs)
+
+        if isinstance(outputs, tuple):
+            output_list = list(outputs)
+        elif isinstance(outputs, Variable) or torch.is_tensor(outputs):
+            output_list = [outputs]
+        out_grads = [grad for grad in args]
+        torch.autograd.backward(output_list, out_grads)
+
+        input_grads = None
+        if isinstance(inputs, tuple):
+            input_grads = tuple(inp.grad for inp in inputs)
+            return (None,) + input_grads
+        elif isinstance(inputs, Variable) or torch.is_tensor(inputs):
+            input_grads = inputs.grad
+            return None, input_grads
+
+
+def checkpoint(run_function, *args):
+    r"""Checkpoint a model or part of the model
+
+    Checkpoint works by trading compute for memory. It can be applied on any
+    part of the model. In the forward pass, the model is run in volatile
+    manner i.e. the activations are not stored. The forward pass save the
+    inputs tuple and the run_function parameter. In the backwards pass, the
+    saved inputs and run_function is retreived, and the forward pass is done
+    on the model again (non-volatile this time) since we need to get the
+    activations values for calculating the gradient and then the gradients are
+    calculated.
+
+    Args:
+        run_function : describes what to run in the forward pass of the model or
+                       part of the model. It should also know how to handle
+                       the inputs passed as the tuple. For example, in LSTM,
+                       user passes (activation, hidden), run_function should
+                       correctly use first input as activation and second input
+                       as hidden
+        args:         tuple containing inputs to the run_function
+
+    Returns:
+        Output of running the run_function on *args
+    """
+    return CheckpointFunction.apply(run_function, *args)
+
+
+def checkpoint_sequential(modules, segments, run_function, *inputs):
```

Comment:
BTW do we really need `run_function`? If it's a sequential, then we can just run `modules[start:end]` right?

Reply:
I wanted to have consistency with the checkpoint() call and also left it to users to determine how their inputs_tuple should be handled and what they want the checkpoint call to run.

Reply:
Well since you're explicitly saying it's sequential, and you're not saying exactly what the strategy for checkpointing is in the docs, then I can't think of many things that people might want to do with a custom closure. I think it would be simpler to remove it, or at least make it optional.

Reply:
I can update the docs for sure and I can move the wrapper for creating run_function here in checkpoint_sequential call()